modules/core/src/test/scala/doobie/util/UpdateLogSuite.scala

```diff
@@ -18,27 +18,29 @@ class UpdateLogSuite extends munit.FunSuite {
 
   import cats.effect.unsafe.implicits.global
 
-  val ioLocal: IOLocal[LogEvent] =
-    IOLocal[LogEvent](null).unsafeRunSync()
+  val logEventRef: Ref[IO, LogEvent] =
+    Ref.of[IO, LogEvent](null).unsafeRunSync()
```

Comment:
IOLocal is no longer works in logHandler because there is `cancelable` which forks a fiber in a middle of query processing

Reply:
Shall we guarantee that IOLocal works with logHandler ?

Reply:
Ouch yes we probably need to. One of our docuemented examples also uses IOLocal to pass additional context to LogHandler.
In a perfect world, I do feel like `IO.cancelable` implementation should pass on the IOLocal though..

Reply:
> In a perfect world, I do feel like `IO.cancelable` implementation should pass on the IOLocal though..

In general, you cannot make these kinds of assumptions when using `IOLocal` with combinators. Further reading:
- https://github.com/typelevel/cats-effect/issues/3100
- https://github.com/typelevel/cats-effect/issues/3385

Reply:
I think there are a lot of problems with maintaining this feature:
1) It is not very functional. I think the most appropriate way to do this - use writer monads, and I suppose that is possible in current setup.
2) It restricts ours capabilities to rely on CE3 functional combinators.

To be honest, I think that IO should be used exclusively in the internal code of an application (or library). The API of a functional library should not provide (or at least guarantee that it will work fine) the use of this tool to users

Reply:
@jatcwang 

Please give a reference to place in the documentation where IOLocal is used. IOLocal can be used to pass context to LogHandler. But It can't be used to modify this context by logHandler (in tests loghandler modifies IOLocal context). I've found this example in doc:

```scala mdoc
import cats.effect.{IOLocal, Ref}
import doobie.util.log.Success

def users = List.range(0, 4).map(n => s"user-$n")

def program: IO[List[String]] =
  for {
    // define an IOLocal where we store the user which caused the query to be run  
    currentUser <- IOLocal("")
    // store all successful sql here, for all users
    successLogsRef <- Ref[IO].of(List.empty[String])
    xa = Transactor.fromDriverManager[IO](
        driver = "org.h2.Driver",
        url = "jdbc:h2:mem:queryspec;DB_CLOSE_DELAY=-1",
        user = "sa", 
        password = "",
        logHandler = Some(new LogHandler[IO] {
          def run(logEvent: LogEvent): IO[Unit] =
            currentUser.get.flatMap(user => successLogsRef.update(logs => s"sql for user $user: '${logEvent.sql}'" :: logs))
        })
      )
    // run a bunch of queries
    _ <- users.parTraverse(user =>
      for {
        _ <- currentUser.set(user)
        _ <- sql"select 1".query[Int].unique.transact(xa)
      } yield ()
    )
    // return collected log messages
    logs <- successLogsRef.get
  } yield logs

program.unsafeRunSync().sorted
```

It will work because child fibers inherit context of parents. So if we set up context in parrent fiber it can be accessed in loghandler. But if we modify context in loghandler this modification can't be propagated to a parent fiber. In tests there is modification of context in child fiber

Reply:
Consider these two examples:

```
import cats.effect.{IO, IOApp, IOLocal}

object IOCancelTest extends IOApp.Simple {

  def childTask(local: IOLocal[Int]): IO[Unit] =
    for {
      _ <- local.get.flatTap(IO.println) // 42
    } yield ()

  override def run: IO[Unit] = for {
    local <- IOLocal(42)
    _ <- childTask(local).cancelable(IO(println("some finalizer")))
  } yield ()
}
```

```
import cats.effect.{IO, IOApp, IOLocal}

object IOCancelTest2 extends IOApp.Simple {

  def childTask(local: IOLocal[Int]): IO[Unit] =
    for {
      _ <- local.set(64)
    } yield ()

  override def run: IO[Unit] = for {
    local <- IOLocal(42)
    _ <- childTask(local).cancelable(IO(println("some finalizer")))
    _ <- local.get.flatTap(IO.println) // 42
  } yield ()
}
```

Reply:
Yep you've found the example in the docs :+1: 

I think trying to modify IOLocal within a ConnectionIO should be quite rare. Since this is a general issue with IOLocal propagation it's not something we can solve here. 

Thanks for updating the tests to use Ref :+1: